pkg/controllers/nsxserviceaccount/nsxserviceaccount_controller.go

```diff
@@ -0,0 +1,242 @@
+		log.Error(err, "unable to fetch NSXServiceAccount CR", "req", req.NamespacedName)
+		return ResultNormal, client.IgnoreNotFound(err)
+	}
+
+	// Since NSXServiceAccount service can only be activated from NSX 4.1.0 onwards,
+	// So need to check NSX version before starting NSXServiceAccount reconcile
+	if !r.Service.NSXClient.NSXCheckVersionForNSXServiceAccount() {
+		err := errors.New("NSX version check failed, NSXServiceAccount feature is not supported")
+		updateFail(r, &ctx, obj, &err)
+		// if NSX version check fails, it will be put back to reconcile queue and be reconciled after 5 minutes
+		return ResultRequeueAfter5mins, nil
+	}
+
+	if obj.ObjectMeta.DeletionTimestamp.IsZero() {
+		metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerUpdateTotal, MetricResType)
+		if !controllerutil.ContainsFinalizer(obj, servicecommon.NSXServiceAccountFinalizerName) {
+			controllerutil.AddFinalizer(obj, servicecommon.NSXServiceAccountFinalizerName)
+			if err := r.Client.Update(ctx, obj); err != nil {
+				log.Error(err, "add finalizer", "nsxserviceaccount", req.NamespacedName)
+				updateFail(r, &ctx, obj, &err)
+				return ResultRequeue, err
+			}
+			log.V(1).Info("added finalizer on CR", "nsxserviceaccount", req.NamespacedName)
+		}
+
+		if obj.Status.Phase == nsxvmwarecomv1alpha1.NSXServiceAccountPhaseRealized {
+			return ResultNormal, nil
+		}
+		if err := r.Service.CreateOrUpdateNSXServiceAccount(ctx, obj); err != nil {
+			if errors.As(err, &nsxutil.RestrictionError{}) {
+				log.Error(err, err.Error(), "nsxserviceaccount", req.NamespacedName)
+				updateFail(r, &ctx, obj, &err)
+				return ResultNormal, nil
+			}
+			log.Error(err, "operate failed, would retry exponentially", "nsxserviceaccount", req.NamespacedName)
+			updateFail(r, &ctx, obj, &err)
+			return ResultRequeue, err
+		}
+		updateSuccess(r, &ctx, obj)
+	} else {
+		if controllerutil.ContainsFinalizer(obj, servicecommon.NSXServiceAccountFinalizerName) {
+			metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerDeleteTotal, MetricResType)
+			if err := r.Service.DeleteNSXServiceAccount(ctx, types.NamespacedName{
+				Namespace: obj.Namespace,
+				Name:      obj.Name,
+			}); err != nil {
+				log.Error(err, "deleting failed, would retry exponentially", "nsxserviceaccount", req.NamespacedName)
+				deleteFail(r, &ctx, obj, &err)
+				return ResultRequeue, err
+			}
+			controllerutil.RemoveFinalizer(obj, servicecommon.NSXServiceAccountFinalizerName)
+			if err := r.Client.Update(ctx, obj); err != nil {
+				log.Error(err, "removing finalizer failed, would retry exponentially", "nsxserviceaccount", req.NamespacedName)
+				deleteFail(r, &ctx, obj, &err)
+				return ResultRequeue, err
+			}
+			log.V(1).Info("removed finalizer", "nsxserviceaccount", req.NamespacedName)
+			deleteSuccess(r, &ctx, obj)
+		} else {
+			// only print a message because it's not a normal case
+			log.Info("finalizers cannot be recognized", "nsxserviceaccount", req.NamespacedName)
```

Comment:
for abnormal case, shall we use error instead of info?

Reply:
In securitypolicy controller, we already use log.info for missing finalizer case. I think we should keep the behavior same.
https://github.com/vmware-tanzu/nsx-operator/blob/32e9730b482c58d4706b7b7e6998d0ae0de6bd89/pkg/controllers/securitypolicy/securitypolicy_controller.go#L155

Reply:
OK